backstory/src/test/java/com/google/sps/perspective/PerspectiveAPIClientTest.java

```diff
@@ -0,0 +1,191 @@
+   * object with the same text as passed in.
+   */
+  @Test
+  public void checkTextSame() {
+    PerspectiveValues values = client.analyze(desiredTypes, "foo");
+
+    Assert.assertEquals("foo", values.getText());
+  }
+
+  /**
+   * Check that analyze() returns a PerspectiveValues
+   * object with the right scores.
+   */
+  @Test
+  public void checkRightScores() {
+    // a Map to be used as the desired scores for the mock API to return 
+    Map desiredScores = new HashMap<AttributeType, Float>();
+
+    // add each of the desiredTypes and a random float between 0 & 1 as their score 
+    for (AttributeType type: desiredTypes) {
+      Random rand = new Random();
+     
+      desiredScores.put(type, rand.nextFloat());
+    }
+
+    mockAPI = createMockAPI(desiredScores);
+    client = new PerspectiveAPIClient(mockAPI);
+    
+    // get the PerspectiveValues object and specifically the output analyses (scores from PerspectiveAPI)
+    PerspectiveValues values = client.analyze(desiredTypes, "foo");
+    Map<AttributeType, Float> outputScores = values.getAttributeTypesToScores();
+
+    // check all the right scores were returned from analyze()
+    for (AttributeType type: desiredTypes) {
+      Assert.assertEquals((float) desiredScores.get(type), outputScores.get(type), 0);
+    }
+  }
+
+  /** 
+   * Returns a mock version of the PerspectiveAPI which will return the scores
+   * passed in through desiredScores or which returns -1 when there is no desired score
+   * (i.e. the map does not contain that attribute type as a key).
+   *
+   * @param desiredScores a map of the scores you want the mockAPI to return 
+   *    (if you don't set any, the mock version of this API will always return -1)
+   * @return the mocked version of the PerspectiveAPI
+   */
+  private static PerspectiveAPI createMockAPI(Map<AttributeType, Float> desiredScores) {
+    // instantiate all the necessary mock objects
+    PerspectiveAPI mockAPI = mock(PerspectiveAPI.class);
+    AnalyzeCommentResponse mockResponse = mock(AnalyzeCommentResponse.class);
+
+    // when analyze is called for our mockAPI, return the mockResponse
+    when(mockAPI.analyze(any(AnalyzeCommentRequest.class))).thenReturn(mockResponse);
+
+    // when getAttributeScore is called we need to be able to grab the type it is looking for
+    // then, check if this AttributeType is in the map and if it is, grab its value, and
+    // set up the next few calls so that this value will be returned. if it isn't in the map,
+    // set up the next few calls so -1 will be returned (-1 signifies no particular value was desired).
+    when(mockResponse.getAttributeScore(any(AttributeType.class))).thenAnswer(new Answer() {
+      public Object answer(InvocationOnMock invocation) {
+        Object[] args = invocation.getArguments();
+        Object mock = invocation.getMock();
+
+        AttributeType type = (AttributeType) args[0];
+
+        float scoreValue = -1;
```

Comment:
How about setting this to a private constant, so that it seems less like a "magic number" (reader is a bit more aware what the significance of this value is)?

```
private static final DEFAULT_ATTRIBUTE_SCORE = -1;
```

Reply:
That's a great idea. Will definitely make code more comprehensible.

Reply:
Done